pysr/julia_helpers.py

```diff
@@ -1,284 +1,34 @@
-
-    if not julia_initialized:
-        _check_for_conflicting_libraries()
-
-    if julia_kwargs is None:
-        julia_kwargs = {"optimize": 3}
-
-    from julia.core import JuliaInfo, UnsupportedPythonError
-
-    _julia_version_assertion()
-    processed_julia_project, is_shared = _process_julia_project(julia_project)
-    _set_julia_project_env(processed_julia_project, is_shared)
-
-    try:
-        info = JuliaInfo.load(julia="julia")
-    except FileNotFoundError:
-        env_path = os.environ["PATH"]
-        raise FileNotFoundError(
-            f"Julia is not installed in your PATH. Please install Julia and add it to your PATH.\n\nCurrent PATH: {env_path}",
-        )
-
-    if not info.is_pycall_built():
-        raise ImportError(_import_error())
-
-    from julia.core import Julia
-
-    try:
-        Julia(**julia_kwargs)
-    except UnsupportedPythonError:
-        # Static python binary, so we turn off pre-compiled modules.
-        julia_kwargs = {**julia_kwargs, "compiled_modules": False}
-        Julia(**julia_kwargs)
-        warnings.warn(
-            "Your system's Python library is static (e.g., conda), so precompilation will be turned off. For a dynamic library, try using `pyenv` and installing with `--enable-shared`: https://github.com/pyenv/pyenv/blob/master/plugins/python-build/README.md#building-with---enable-shared."
-        )
-
-    using_compiled_modules = (not "compiled_modules" in julia_kwargs) or julia_kwargs[
-        "compiled_modules"
-    ]
-
-    from julia import Main as _Main
-
-    Main = _Main
-
-    if julia_activated_env is None:
-        julia_activated_env = processed_julia_project
 
-    if julia_initialized and julia_kwargs_at_initialization is not None:
-        # Check if the kwargs are the same as the previous initialization
-        init_set = set(julia_kwargs_at_initialization.items())
-        new_set = set(julia_kwargs.items())
-        set_diff = new_set - init_set
-        # Remove the `compiled_modules` key, since it is not a user-specified kwarg:
-        set_diff = {k: v for k, v in set_diff if k != "compiled_modules"}
-        if len(set_diff) > 0:
-            warnings.warn(
-                "Julia has already started. The new Julia options "
-                + str(set_diff)
-                + " will be ignored."
-            )
-
-    if julia_initialized and julia_activated_env != processed_julia_project:
-        Main.eval("using Pkg")
-
-        io_arg = _get_io_arg(quiet)
-        # Can't pass IO to Julia call as it evaluates to PyObject, so just directly
-        # use Main.eval:
-        Main.eval(
-            f'Pkg.activate("{_escape_filename(processed_julia_project)}",'
-            f"shared = Bool({int(is_shared)}), "
-            f"{io_arg})"
-        )
-
-        julia_activated_env = processed_julia_project
+# Required to avoid segfaults (https://juliapy.github.io/PythonCall.jl/dev/faq/)
+if os.environ.get("PYTHON_JULIACALL_HANDLE_SIGNALS", "yes") != "yes":
```

Comment:
@cjdoris i think this might address your other comment? I didn’t do the same for threads but I can add that as well.

Reply:
Yes I think using warnings to advise the user of the intended settings is a good idea. Setting these in the package is a bad idea in my opinion because you'll get very different behaviour depending on the order modules are imported.

Relatedly, I've added a warning in JuliaCall - if you start Julia with multiple threads but don't set `PYTHON_JULIACALL_HANDLE_SIGNALS` then you get a warning advising to set this env var.

Reply:
> Setting these in the package is a bad idea in my opinion because you'll get very different behaviour depending on the order modules are imported.

I don’t disagree, but asking a user to set an environment variable to get PySR basics working is not a good user experience. Some of my users don’t even know what an environment variable is. Everything needs to be pretty automatic to attract regular scientists (my target users) — I’ve even received requests to have a downloadable app with a GUI (no code at all!).

Reply:
Absolutely - keep it in if it's best for your users, I'm just encouraging best practice.

I wonder if it would be a good idea to be able to set these options in `juliapkg.json` to override the default behaviour. This would allow different packages to cooperate on the values, and works even when JuliaCall is already imported before your own package.

Reply:
> This would allow different packages to cooperate on the values, and works even when JuliaCall is already imported before your own package.

That’s a great idea!